hy/compiler.py

```diff
@@ -776,14 +714,42 @@ def hy_eval(
         import_stdlib=import_stdlib,
     )
 
-    if ast_callback:
-        ast_callback(_ast, expr)
+    if globals is None:
+        globals = module.__dict__
 
     # Two-step eval: eval() the body of the exec call
-    eval(ast_compile(_ast, filename, "exec"), module.__dict__, locals)
+    eval(ast_compile(_ast, filename, "exec"), globals, locals)
 
     # Then eval the expression context and return that
-    return eval(ast_compile(expr, filename, "eval"), module.__dict__, locals)
+    return eval(ast_compile(expr, filename, "eval"), globals, locals)
+
+
+def hy_eval_user(model, globals = None, locals = None, module = None):
+    # This function is advertised as `hy.eval`.
+    """An equivalent of Python's :func:`eval` for evaluating Hy code. The chief difference is that the first argument should be a :ref:`model <models>` rather than source text. If you have a string of source text you want to evaluate, convert it to a model first with :hy:func:`hy.read` or :hy:func:`hy.read-many`::
+
+        (hy.eval '(+ 1 1))             ; => 2
+        (hy.eval (hy.read "(+ 1 1)"))  ; => 2
+
+    The optional arguments ``globals`` and ``locals`` work as in the case of ``eval``. There's one more optional argument, ``module``, which can be a module object or a string naming a module. The module's ``__dict__`` attribute can fill in for ``globals``, if ``module`` is provided but ``globals`` isn't, but the primary purpose of ``module`` is to control where macro calls are looked up. Without this argument, the calling module of ``hy.eval`` is used instead. ::
+
+        (defmacro my-test-mac [] 3)
+        (hy.eval '(my-test-mac))                 ; => 3
+        (import hyrule)
+        (hy.eval '(my-test-mac) :module hyrule)  ; NameError
+        (hy.eval '(list-n 3 1) :module hyrule)   ; => [1 1 1]"""
+
+    value = hy_eval(
+        hytree = model,
+        locals = (inspect.getargvalues(inspect.stack()[1][0]).locals
```

Comment:
I think this is slightly incorrect; the documentation for both Python's `eval` and `expr` state that if `locals` is not passed but `globals` is, then `globals` is also used for the local vars. Compare:
```
=> (eval "locals().keys()" {})
(dict-keys ["__builtins__"])
=> (hy.eval '(.keys (locals)) {})
(dict-keys [... (all of the local vars) ...])
```

Reply:
Hmm, this condition expression here now looks confusing. Since we have the None-check and assignment for `locals` at 736-737, at first glance it looks like the `if locals is None` on 742 will never be False.

In any case, the `locals`/`globals` thing will already be handled by `eval` anyway, so I think the clearer thing is to reassign `locals` only when `all(arg is None for arg in (locals, globals, module))` (or something more elegant), but otherwise pass `locals` to `hy_eval` unmodified.

Reply:
I'm not 100% sure what change you're suggesting. Could you push a commit?

Reply:
```python
    value = hy_eval(
        hytree = model,
        globals = globals,
        locals = (inspect.getargvalues(inspect.stack()[1][0]).locals
            if all(arg is None for arg in (locals, globals, module)) else locals),
        module = get_compiler_module(module, None, True),
        import_stdlib=False)
```

Reply:
I tried replacing the assignment to `value` thus, and removing the preceding `if locals is None: locals = globals` and the following `if locals: …`, and the result was a test failure `NameError: name 'hy' is not defined` in `tests/native_tests/hy_eval.hy::test_locals`, which seems logical enough.

Reply:
ah here's the full diff, I set `import_stdlib=False` in the snippet above so you need the builtins injection as well for it to work:

```diff
diff --git a/hy/compiler.py b/hy/compiler.py
index af0cc8dc..f61d807a 100755
--- a/hy/compiler.py
+++ b/hy/compiler.py
@@ -1,4 +1,5 @@
 import ast
+import builtins
 import copy
 import importlib
 import inspect
@@ -733,18 +734,16 @@ def hy_eval_user(model, globals = None, locals = None, module = None):
         (hy.eval '(my-test-mac) :module hyrule)  ; NameError
         (hy.eval '(list-n 3 1) :module hyrule)   ; => [1 1 1]"""
 
-    if locals is None:
-        locals = globals
+    if globals is not None and '__builtins__' not in globals:
+        globals['__builtins__'] = dict(vars(builtins), hy=hy)
     value = hy_eval(
         hytree = model,
         globals = globals,
         locals = (inspect.getargvalues(inspect.stack()[1][0]).locals
-            if locals is None and module is None
-            else locals),
-        module = get_compiler_module(module, None, True))
-    if locals:
-        # Remove the implicitly added `hy`.
-        del locals['hy']
+                  if all(arg is None for arg in (locals, globals, module))
+                  else locals),
+        module = get_compiler_module(module, None, True),
+        import_stdlib=False)
     return value
```